java/src/main/java/com/google/appengine/tools/mapreduce/impl/shardedjob/ShardedJobRunner.java

```diff
@@ -413,8 +419,10 @@ private <T extends IncrementalTask> void runAndUpdateTask(
                                                             final int sequenceNumber,
                                                             final ShardedJobStateImpl<T> jobState,
                                                             IncrementalTaskState<T> taskState) {
-    ShardRetryState<T> retryState = null;
-    Transaction postRunUpdate = null;  //txn limited to 60s, so can't open this before run() call
+
+    RetryType retryType = RetryType.NONE;
```

Comment:
reorganized the code, so all the handle  failures happens at the end, and with retries